### PR TITLE
fix criterion.weight in CheckPoint resnet50 #422

### PR DESCRIPTION
**Pull request by kodinkod**
i use :
def on_save_checkpoint(self, trainer: pl.Trainer, pl_module: pl.LightningModule, checkpoint) - Called when saving a checkpoint to give you a chance to store anything else you might want to save. I think it will be convenient to adjust the checkpoint.  

I had difficulties in testing and I have found only such a testing option so far, I will be glad of new knowledge and recommendations

### CODE REVIEW

**Comment by AlekseySh** [comment]
Hey, @kodinkod! Thank you for working on this.

The problem with your solution is that the saving logic is linked with the logic of computing metrics. For example, it means that you cannot save/load checkpoint correctly if you don't calculate metric or if you calculate it differently (using another class).

**Comment by AlekseySh** [comment]
As for the tests, here is a good example of how we test pipelines end-to-end: [link](https://github.com/OML-Team/open-metric-learning/tree/main/tests/test_runs/test_pipelines). But don't work on the test before we find a good solution.

**Comment by AlekseySh** [comment]
I think that in general, it's okay to save the criterion's state, because it may be used later to restore the training. So, probably it's better to change the problematic function.

What if we change this
` assert all(k.startswith(prefix) for k in state_dict.keys())`
to something like this:
`assert all(k.startswith(prefix) for k in state_dict.keys() if not k.startswith('criterion'))`
?

It's a bit hacky, but better logically. 

I'm also open to other suggestions.

**Comment by kodinkod** [comment]
I also thought that this would solve my problem, but with such a hacker solution I get 
```
self.model.load_state_dict(state_dict, strict=True)     
Error(s) in loading state_dict for ResNet:
        Unexpected key(s) in state_dict: "criterion.weight". 
```

`on_save_checkpoint`  just controls what is saved in State_dict;  

1. We can redefine `on_save_checkpoint`  in each metric class. 
2. We can somehow competently load each model by saving criterion.weight, and deleting it in state_dict so that we can use criterion.weight;

Also, I don't quite understand what  saved in `criterion.weight`

**Comment by kodinkod** [comment]
If it is convenient for you, you can also write to me in telegam to discuss the problem: @kodlovecats

**Comment by AlekseySh** [comment]
I think you got that error because `strict=True`, try to make it `False`.


> We can redefine on_save_checkpoint in each metric class.

Unfortunately, we cannot, because the idea is that users can have their own metrics. And, as I said before, probably I don't want to calculate any metric, but still want to save a checkpoint.

> We can somehow competently load each model by saving criterion.weight, and deleting it in state_dict so that we can use criterion.weight;

By the way, is that possible to control saving behavior on `pl.LightningModule` level, [here](https://github.com/OML-Team/open-metric-learning/blob/main/oml/lightning/modules/extractor.py#L14)?

> Also, I don't quite understand what saved in criterion.weight

Check [it](https://github.com/OML-Team/open-metric-learning/blob/a1a5d3e74baf15896b573ce47fd27d1a4b9924b4/oml/losses/arcface.py#L57) out.

**Comment by kodinkod** [comment]
> I think you got that error because strict=True, try to make it False. 
yes, this fixes the problem;

**Comment by AlekseySh** [comment]
I didn't suggest changing `strict` parameter from `True` to `False`, because I'm also worried that it may cause problems. I just wanted to double-check that we really understand what is going on. 

I think the optimal solution is to remove `criterion` key from the dict in all the existing models when we load them from the checkpoints. What do you think?

**Comment by kodinkod** [comment]
It seems that only this method remains;  I think we can implement a function that will remove `criterion ` from the  `state_dict` and call it when loading the model. What do you think ?

**Comment by AlekseySh** [comment]
@kodinkod Yes, I agree with you :) 

Check out this comment one more time before writing tests: https://github.com/OML-Team/open-metric-learning/pull/430#issuecomment-1670101780

**Comment by AlekseySh** [comment]
@kodinkod Hey, are u going to finish this?

**Comment by AlekseySh** [comment]
Just a friendly reminder

**Comment by kodinkod** [comment]
sorry, I was busy with work, I'll try to do everything soon

**Comment by AlekseySh** [comment]
@kodinkod no worries, ping we when the code is ready for review

**Comment by kodinkod** [comment]
is the last commit a good practice? I would like to know this before I start porting similar code to other extractors

File: tests/test_oml/test_checpoints/test_checpoints_dict.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+from pathlib import Path
```
Comment (AlekseySh) [inline tests/test_oml/test_checpoints/test_checpoints_dict.py:1]: I think we don't need it anymore

**Comment by AlekseySh** [comment]
@kodinkod I recommend you to run tests locally in a new env or in docker, so it will be easier to pass CI (check Makefile)

File: tests/test_oml/test_checpoints/test_checpoints_dict.py (hunk 1 above)
Comment (kodinkod) [reply tests/test_oml/test_checpoints/test_checpoints_dict.py:1]: yes

**Comment by kodinkod** [comment]
@AlekseySh   If we're not in a rush I can continue working on it. It remains to fix a few extractors.

**Comment by AlekseySh** [comment]
@kodinkod not in rush for sure, take your time

**Comment by AlekseySh** [comment]
@kodinkod Hey! let's finish this! it shouldn't take too much time :)

**Comment by kodinkod** [comment]
@AlekseySh I'm very sorry, I'm interested in participating in the project. but I've been incredibly busy at university and at work. I'm almost finished working on pr. I'm having difficulty writing tests for vit_clip and clip_unicom. I will be grateful for your help. Now I have little time and I'm afraid to make a mistake in a hurry.

thank you for the opportunity pr. I apologize again :(

**Comment by AlekseySh** [comment]
@kodinkod thank you anyway! text me in telegram for cooperation (@lexibender)
